core/bluetooth/src/main/java/edu/stanford/spezikt/core/bluetooth/BluetoothViewModel.kt

```diff
@@ -0,0 +1,131 @@
+import androidx.core.app.ActivityCompat
+import androidx.lifecycle.ViewModel
+import edu.stanford.spezikt.core.bluetooth.model.BloodPressureMeasurement
+import edu.stanford.spezikt.core.bluetooth.model.Measurement
+import edu.stanford.spezikt.core.bluetooth.model.WeightMeasurement
+import java.util.UUID
+
+class BluetoothViewModel : ViewModel(), BluetoothScanner.DeviceScanListener,
+    DeviceConnectionListener,
+    CustomGattCallback.DataListener {
+    private lateinit var bluetoothManager: BluetoothManager
+    var bluetoothScanner: BluetoothScanner? = null
+    var deviceConnector: DeviceConnector? = null
+
+    private val serviceUuidsToConverters = mapOf( // scale and blood pressure service UUIDs
+        BluetoothUUID(
+            UUID.fromString("0000181d-0000-1000-8000-00805f9b34fb"),
+            UUID.fromString("00002a9d-0000-1000-8000-00805f9b34fb")
+        ) to WeightDataConverter(),
+        BluetoothUUID(
+            UUID.fromString("00001810-0000-1000-8000-00805f9b34fb"),
+            UUID.fromString("00002a35-0000-1000-8000-00805f9b34fb")
+        ) to BloodPressureDataConverter()
+    )
+
+    val connectedDevices = mutableStateOf(listOf<BluetoothDevice>())
+    val bloodPressureData = mutableStateOf<BloodPressureMeasurement?>(null)
+    val weightData = mutableStateOf<WeightMeasurement?>(null)
+
+    val showDialog = mutableStateOf(false)
+    val currentMeasurement = mutableStateOf<Measurement?>(null)
+
+
+    @RequiresApi(Build.VERSION_CODES.S)
+    fun start(context: Context, activity: Activity) {
+        try {
+            Log.d("BluetoothActivity", "onCreate: Content set")
+            bluetoothManager = BluetoothManager(context)
+            requestBluetoothScanPermission(activity)
+
+            if (bluetoothManager.isBluetoothEnabled()) {
+                if (ActivityCompat.checkSelfPermission(
+                        context,
+                        Manifest.permission.BLUETOOTH_CONNECT
+                    ) != PackageManager.PERMISSION_GRANTED
+                ) {
+                    return
+                }
+
+                bluetoothManager.getAdapter()?.let { btAdapter ->
+                    Log.d("BluetoothActivity", "start Scan")
+                    bluetoothScanner = BluetoothScanner(btAdapter, this)
+                    bluetoothScanner?.startScan(serviceUuidsToConverters.keys.map { it.serviceUUID }
+                        .toList())
+                }
+            }
+        } catch (e: Exception) {
+            Log.e("BluetoothActivity", "Error starting bluetooth", e)
+            bluetoothScanner?.stopScan()
+            deviceConnector?.close()
+        }
+    }
+
+    override fun onDeviceFound(result: ScanResult) {
+        try {
+            Log.d("BluetoothActivity", "Device found: ${result.device.address}")
+            if (connectedDevices.value.contains(result.device)) {
+                return
+            }
+            deviceConnector = DeviceConnector(
```

Comment:
Actually, it's the other way around, isn't it? if there is already a connection and the device is found again, we ignore it (or at least that would be what I had intended). But the old connection remains because we currently only need a single connection. We only use the Bluetooth connection to subscribe to a specific service on the device. One connection per device should always be sufficient for this purpose